File: app/src/main/java/com/example/locationsample_kotlin_android/location/LocationLifecycleObserver.java
package com.example.locationsample_kotlin_android.location;

import android.content.Context;
import android.location.Location;
import android.os.Handler;
import androidx.annotation.Nullable;
import androidx.lifecycle.*;

public class LocationLifecycleObserver implements LifecycleObserver, LocationStatusListener {
    private Context mContext;

    private static final int DEFAULT_MAX_LOCATION_REQUEST_TIME = 15000;

    private boolean mIsSingleLocationRequest = false;
    private Handler mLocationRequestTimeoutHandler;

    private LocationProvidersContract mLocationProvidersContract;

    private MutableLiveData<LocationStatus> mLocationResponseLiveData;
    private long mMaxLocationRequestTime = DEFAULT_MAX_LOCATION_REQUEST_TIME;

    /**
     * @param context
     * @param maxLocationRequestTime Sets the max location updates request time
     *                               if exceeded stops the location updates and returns error
     *                               set to Null for default time {#LocationLifecycleObserver.DEFAULT_MAX_LOCATION_REQUEST_TIME}.
     */
    public LocationLifecycleObserver(Context context, @Nullable Long maxLocationRequestTime) {
        this.mContext = context;
        if (maxLocationRequestTime != null)
            mMaxLocationRequestTime = maxLocationRequestTime;
        mLocationResponseLiveData = new MutableLiveData<>();
    }

    public LocationLifecycleObserver singleLocationRequest() {
        mIsSingleLocationRequest = true;
        return this;
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    void onDestroy() {
        stopLocationUpdates();
    }

    /**
     * creates Location Request
     *
     * @return
     */
    public MutableLiveData<LocationStatus> startNetworkLocationUpdates() {
        if (mIsSingleLocationRequest)
            startLocationRequestTimer();
        createNetworkLocationRequest();
        return mLocationResponseLiveData;
    }

    /**
     * Creates location updates
     *
     * @param interval
     * @param fastestInterval
     * @return
     */
    public MutableLiveData<LocationStatus> startFusedLocationUpdates(long interval, long fastestInterval) {
        if (mIsSingleLocationRequest)
            startLocationRequestTimer();
        createFusedLocationRequest(interval, fastestInterval);
        return mLocationResponseLiveData;
    }

    @Override
    public void onLocationRetrieved(Location location) {
        if (mIsSingleLocationRequest)
            stopLocationUpdates();
        mLocationResponseLiveData.setValue(LocationStatus.success(location));
    }

    @Override
    public void onLocationRetrieveError(LocationStatus locationStatus) {
        if (mIsSingleLocationRequest)
            stopLocationUpdates();
        mLocationResponseLiveData.setValue(locationStatus);
    }

    /**
     * Starts a timer to stop the request of the location updates after 30 seconds
     */
    private void startLocationRequestTimer() {
        mLocationRequestTimeoutHandler = new Handler();
        mLocationRequestTimeoutHandler.postDelayed(runnable, mMaxLocationRequestTime);
    }

    private Runnable runnable = new Runnable() {
        @Override
        public void run() {
            stopLocationUpdates();

            mLocationProvidersContract.fetchLatestKnownLocation();
        }
    };

    public void setSingleLocationRequest(boolean isSingleLocationRequest) {
        mIsSingleLocationRequest = isSingleLocationRequest;
    }

    private void createFusedLocationRequest(long interval, long fastestInterval) {
        mLocationProvidersContract = new LocationProviderFused(mContext, this, interval, fastestInterval);
        mLocationProvidersContract.requestLocationUpdates();
    }

    private void createNetworkLocationRequest() {
        mLocationProvidersContract = new LocationProviderNetwork(mContext, this);
        mLocationProvidersContract.requestLocationUpdates();
    }

    public void stopLocationUpdates() {
        if (mLocationProvidersContract != null)
            mLocationProvidersContract.stopLocationUpdates();

        if (mLocationRequestTimeoutHandler != null && runnable != null)
            mLocationRequestTimeoutHandler.removeCallbacks(runnable);
    }
}

File: app/src/main/java/com/example/locationsample_kotlin_android/location/LocationLifeCycleObserver.kt
package com.example.locationsample_kotlin_android.location

import android.content.Context
import android.location.Location
import android.os.Handler
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.OnLifecycleEvent
import com.example.locationsample_kotlin_android.location.location_providers.LocationOptions
import com.example.locationsample_kotlin_android.location.location_providers.LocationProviders
import com.example.locationsample_kotlin_android.location.location_providers.LocationProvidersTypes
import com.example.locationsample_kotlin_android.location.location_providers.LocationStatusListener
import com.example.locationsample_kotlin_android.location.location_providers.fused.FusedLocationOptions
import com.example.locationsample_kotlin_android.location.location_providers.fused.FusedLocationProvider
import com.example.locationsample_kotlin_android.location.location_providers.network.NetworkLocationOptions
import com.example.locationsample_kotlin_android.location.location_providers.network.NetworkLocationProvider

/**
 * todo
 * @param mContext
 * @param mMaxLocationRequestTime Sets the max location updates request time
 *      if exceeded stops the location updates and returns error <P>
 *      @Default{#LocationLifecycleObserver.DEFAULT_MAX_LOCATION_REQUEST_TIME}.
 * @param mIsSingleLocationRequest todo
 */
class LocationLifeCycleObserver(lifecycle: Lifecycle, private val mContext: Context,
                                private var mMaxLocationRequestTime: Long = Constants.DEFAULT_MAX_LOCATION_REQUEST_TIME,
                                private var mIsSingleLocationRequest: Boolean = false) : LifecycleObserver {

    private lateinit var mLocationProviders: LocationProviders
    private val mLocationResponseLiveData: MutableLiveData<LocationStatus> = MutableLiveData()
    private var mLocationRequestTimeoutHandler: Handler = Handler()

    init {
        lifecycle.addObserver(this)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroy() {
        stopLocationUpdates()
    }

    fun requestLocationUpdates(locationProvidersTypes: LocationProvidersTypes, locationOptions: LocationOptions): MutableLiveData<LocationStatus> {
        return when (locationProvidersTypes) {
            LocationProvidersTypes.FUSED_LOCATION_PROVIDER -> {
                if (locationOptions !is FusedLocationOptions) {
                    throw Exception("Fused location provider options not found should be [FusedLocationOptions]")
                }
                startFusedLocationUpdates(locationOptions)
            }
            LocationProvidersTypes.NETWORK_LOCATION_PROVIDER -> {
                if (locationOptions !is NetworkLocationOptions) {
                    throw Exception("Network location provider options not found should be [NetworkLocationOptions]")
                }
                startNetworkLocationUpdates(locationOptions)
            }
        }
    }

    /**
     * creates Location Request.
     *
     * @return
     */
    private fun startNetworkLocationUpdates(networkLocationOptions: NetworkLocationOptions): MutableLiveData<LocationStatus> {
        if (mIsSingleLocationRequest)
            startLocationRequestTimer()
        requestNetworkLocationUpdates(networkLocationOptions)
        return mLocationResponseLiveData
    }

    /**
     * Creates location updates.
     *
     * @param fusedLocationOptions
     * @return
     */
    private fun startFusedLocationUpdates(fusedLocationOptions: FusedLocationOptions): MutableLiveData<LocationStatus> {
        if (mIsSingleLocationRequest)
            startLocationRequestTimer()
        requestFusedLocationUpdates(fusedLocationOptions)
        return mLocationResponseLiveData
    }

    private val mLocationStatusListener: LocationStatusListener = object : LocationStatusListener {
        override fun onLocationRetrieved(location: Location) {
            if (mIsSingleLocationRequest)
                stopLocationUpdates()
            mLocationResponseLiveData.value = LocationStatus.Success(location)
        }

        override fun onLocationRetrieveError(locationStatus: LocationStatus?) {
            if (mIsSingleLocationRequest)
                stopLocationUpdates()
            mLocationResponseLiveData.value = locationStatus
        }
    }

    private fun requestFusedLocationUpdates(fusedLocationOptions: FusedLocationOptions) {
        mLocationProviders = FusedLocationProvider(mContext, fusedLocationOptions)
        mLocationProviders.requestLocationUpdates(mLocationStatusListener)
    }

    private fun requestNetworkLocationUpdates(networkLocationOptions: NetworkLocationOptions) {
        mLocationProviders = NetworkLocationProvider(mContext, networkLocationOptions)
        mLocationProviders.requestLocationUpdates(mLocationStatusListener)
    }

    fun stopLocationUpdates() {
        if (::mLocationProviders.isInitialized)
            mLocationProviders.stopLocationUpdates()
        mLocationRequestTimeoutHandler.removeCallbacks(runnable)
    }

    /**
     * Starts a timer to stop the request of the location updates after [mMaxLocationRequestTime] seconds
     */
    private fun startLocationRequestTimer() {
        mLocationRequestTimeoutHandler.postDelayed(runnable, mMaxLocationRequestTime)
    }

    private val runnable: Runnable = Runnable {
        stopLocationUpdates()
        mLocationProviders.fetchLatestKnownLocation()
    }
}